lib/ndr_import/xml/table.rb

```diff
@@ -24,28 +31,114 @@ def footer_lines
+        # Start with a fresh set of @augmented_columns for each line, adding new mappings as
+        # required for each `line`
+        @augmented_columns       = @columns.deep_dup
+        @augmented_column_xpaths = column_xpaths.deep_dup
+
+        unmapped_nodes(line).each do |unmapped_node|
+          existing_column = find_existing_column_for(unmapped_node.dup)
+          next unless existing_column
+
+          unmapped_node_parts   = unmapped_node_parts(unmapped_node)
+          klass_increment_match = unmapped_node.match(/\[(\d+)\]/)
+          raise "could not identify klass for #{unmapped_node}" unless klass_increment_match
+
+          new_column = NdrImport::Xml::ColumnMapping.new(existing_column, unmapped_node_parts,
+                                                         klass_increment_match[1], line,
+                                                         @klass).call
+          @augmented_columns << new_column
+          @augmented_column_xpaths << build_xpath_from(new_column)
+        end
+      end
+
+      def xml_line_from(line)
+        @augmented_column_xpaths.map do |column_xpath|
+          # Augmenting the column mappings should account for repeating sections/items
+          # TODO: Is this needed now that we removed "duplicated" klass mappings?
+          line.xpath(column_xpath).count > 1 ? '' : line.xpath(column_xpath).inner_text
+        end
+      end
+
+      def find_existing_column_for(unmapped_node)
+        # Remove any e.g. [2] which will be present on repeating sections
+        unmapped_node.gsub!(/\[\d+\]/, '')
+        unmapped_node_parts = unmapped_node_parts(unmapped_node)
+        columns.detect do |column|
+          column['column'] == unmapped_node_parts[:column_name] &&
+            column.dig('xml_cell', 'relative_path') == unmapped_node_parts[:column_relative_path] &&
+            column.dig('xml_cell', 'attribute') == unmapped_node_parts[:column_attribute]
+        end
+      end
+
+      def unmapped_node_parts(unmapped_node)
+        unmapped_node_parts       = unmapped_node.split('/')
+        unmapped_column_attribute = new_column_attribute_from(unmapped_node_parts)
+
+        { column_attribute: unmapped_column_attribute,
+          column_name: new_column_name_from(unmapped_node_parts, unmapped_column_attribute),
+          column_relative_path: new_relative_path_from(unmapped_node_parts,
+                                                       unmapped_column_attribute) }
+      end
+
+      def new_column_attribute_from(unmapped_node_parts)
+        unmapped_node_parts.last.starts_with?('@') ? unmapped_node_parts.last[1...] : nil
+      end
+
+      def new_column_name_from(unmapped_node_parts, unmapped_column_attribute)
+        unmapped_column_attribute.present? ? unmapped_node_parts[-2] : unmapped_node_parts.last
+      end
+
+      def new_relative_path_from(unmapped_node_parts, unmapped_column_attribute)
+        upper_limit = unmapped_column_attribute.present? ? -3 : -2
```

Comment:
Maybe add a developer comment, explaining why the upper limit is -3 or -2. I've read all the code, and I have no idea why. I _think_ `unmapped_node_parts` is generally an array, so this dropping either the last 2 or 1 elements respectively? Maybe a developer comment explaining the structure of `unmapped_node_parts`?

Reply:
Agreed, I'll add some notes. Brief explanation:

xpaths are e.g. `Record/Demographics/Sex/@code` or `Record/Demographics/Surname`

For the relative path, we want `Record/Demographics`, the `upper_limit` removes the "field name" (`Sex` or `Surname` here) and optionally the attribute (`@code` here) if present.

Reply:
Comment added in b45119a6d95f8cc357e550d84856a539826b85fa